### ISSUE TITLE
How to apply data transformation when using skorch with pytorch dataloader

### ISSUE THREAD

**Issue by xiaolongwu0713** [skorch-dev/skorch#756]
Hi, I have a special case.
I have 1000 time series samples. Each sample is a 2D data of 10 feature lasting for 100 seconds at 1 datapoint per second. 
So I have 1000 * 10 *100 (means: samples * feature * time) raw data. 
For each sample data of size 10*100 , I need to use a window size of 5 to slide slide_times times along the time dimension. Then I am going to predict a scalar from this 2D (10 feature* 5 wind_size) window. So now one 2D sample becomes a 3D of size (10feature * 5 wind_size * slide_times) and the task is to predict a vector of slide_times  long. 
Before Skorch, I just use dataloader of batch_size=1 to load one sample, then use a window to slide this sample to make my 3D data. In this case, the slide_times is my real batch size in the following net even though the dataloader class has a batch_size=1.
Now come to the skorch. 
If I use the same dataloader as before and put it to skorch, the first axis batch_size would be 1 because I do sample one 2D sample a time and convert to 3D. However 1 is not my actual batch_size, slide_times is my batch size. 
So here comes the question: how can I squeeze out the first axis of number 1 like what I did before? where should apply this in skorch env?

**Comment by BenjaminBossan** [comment]
Did you try whether it works or not? I.e. set `batch_size=1` and pass your custom data loader to the skorch net (`iterator_train` and `iterator_valid`arguments)?

**Comment by xiaolongwu0713** [comment]
Benjamin, I can't just do that. I have to reorganize my data by sliding window through the 2D sample.  So I am try to use regular dataloader to give me a 2D sample, slide a window to convert to 3D then feed into skorch.

**Comment by BenjaminBossan** [comment]
> I have to reorganize my data by sliding window through the 2D sample.

Where exactly does that happen? I thought inside the data loader, but it sounds like you perform this after passing the data through the data loader.

Overall, your situation is a bit too abstract for me to give concrete help. Any kind of code that shows how you're currently doing it would help a lot.

**Comment by xiaolongwu0713** [comment]
Yes, I understand my discribution is very abstract. Check the example code below:
 
First, I convert 2D sample to 3D in dataset class by sliding a window through it:

# simple dataset
class SEEGDataset(Dataset):
    # x_tensor: (sample, channel, datapoint(feature)) type = torch.tensor
    # y_tensor: (sample,) type = torch.tensor
    def __init__(self, x_tensor, y_tensor):
        self.x = x_tensor
        self.y = y_tensor
        assert self.x.shape[0] == self.y.shape[0]
    def __getitem__(self, index):
        return self.x[index], self.y[index]
    def __len__(self):
        return self.y.shape[0]

# here trainx has dimension of: (samples, features, times)
dataset_train = SEEGDataset(trainx, trainy)
train_loader = DataLoader(dataset=dataset_train, batch_size=1, shuffle=True, pin_memory=False)

for epoch in range(epochs):
    for trial, (trainx, trainy) in enumerate(train_loader): 
                # here I have to use a window to slide through time dimension.
                for bs in range(slide_times):
                    x[bs, 0, :, :] = trainx[0, :, bs*step:(bs*step + T)]
                    target[bs, 0] = trainy[0, bs * step + T + 1]
                y_pred = net(torch.from_numpy(x).float())

Now I am thinking if I can use a pipline to do the trick.  Here is the idea.
I use the dataloader like I did before. Then I put the window-sliding operation inside the pipline, which would be like below:

pipe = Pipeline([
    ('wind_slide', wind_slide()),
    ('net', net),
])

pipe.fit(X, y)
Please advise me if this is the proper way.

**Comment by xiaolongwu0713** [comment]
or can I just put the transform operation inside the net work forward call? like below:
def forward(self, x):  # ([128, 1, 4, 1024]): (batch_size, )
        # convert 2D to 3D by sliding a window
        x=windTo3D(x,self.wind,self.step)
        .....

**Comment by BenjaminBossan** [comment]
Thanks, I now have a better understanding.

> x[bs, 0, :, :] = trainx[0, :, bs*step:(bs*step + T)]
> target[bs, 0] = trainy[0, bs * step + T + 1]

Where do the `x` and `target` come from?

> Then I put the window-sliding operation inside the pipline

It depends. The pipeline transformation is applied exactly once on your data. If that is okay for you (you have enough memory etc.), you could do it like that. If the transformation is expensive, it might be even better to just transform your data like that once and store it in that fashion for future use.

A reason not to do this is that sklearn `Pipeline`s usually don't expect the number of samples to change, I'm not sure if your transformation does that or not. Also, I don't know exactly what your `trainy` is and if it needs to be transformed as well. If so, it will not work in a pipeline, since the pipeline leaves `y` untouched.

> or can I just put the transform operation inside the net work forward call?

Technically, that is possible, but overall might not be a good idea because you would be mixing the data transformation logic with the neural net logic. It sounds like it could cause bugs down the line. Moreover, if you need to transform `y` as well, I wouldn't put it inside your `Module`, since that is not where the loss is calculated.

Here is another suggestion: Put the transformation inside the dataset. It could look roughly like this:

```python
class SEEGDataset(Dataset):
    ...
    def __getitem__(self, index):
        xb, yb = self.x[index], self.y[index]
        xb = windTo3D(xb)
        # transform yb as well if necessary
        return xb, yb
```

Since I don't understand everything that's going on, there could be a reason why this will not work, but at least to me, this seems like the right place to put it. Why? First of all, the data transformation should be logically tied to the dataset. Second, if the transform is costly in terms of memory, this way you don't need to apply it on all your data at once (but you have to run it multiple times, so you trade off memory for speed). And third, this should allow you to use a batch size > 1, which can help with runtime performance.

**Comment by xiaolongwu0713** [comment]
Really great post. I think it's a good suggestion. I agree it's not a best practice to do data transformation in module, so I put all my transformation inside the dataset class.
